lib/einhorn/config_file.rb

```diff
@@ -0,0 +1,182 @@
+      bind_fds = []
+      bind = desired.map do |(addr, port), conf|
+        bind_fds << conf[:fd]
+        [addr, port, conf[:flags]]
+      end
+      Einhorn::State.bind = bind
+      Einhorn::State.bind_fds = bind_fds
+    end
+
+    def self.apply_bind(addr, port, config, desired_config, current)
+      if !desired_config
+        # No one wants this addr/port anymore, let's shut it down!
+        fd = config[:fd]
+        Einhorn.log_info("Closing fd for #{addr}:#{port} (#{fd})")
+        IO.new(fd).close
+        return
+      end
+
+      # Make sure that the desired FD is available
+      desired_config[:fd] ||= config[:fd]
+      desired_fd = desired_config[:fd]
+      desired_flags = desired_config[:flags]
+
+      # Check on whether anyone's squatting on our FD
+      holder_addr, holder_conf = current.detect {|k, conf| conf[:fd] == desired_fd}
+      if holder_addr && holder_addr == [addr, port]
+        # No binding needed here
+      elsif holder_addr
+        # Someone else is occupying this FD. Need to dup their FD and
+        # clear up the current slot.
+        holder_io = IO.new(desired_fd)
+
+        fd2 = holder_io.fcntl(Fcntl::F_DUPFD)
+        Einhorn.log_info("Clearing room for #{addr}:#{port} by renumbering fd for #{holder_addr.join(':')} from #{desired_fd} -> #{fd2}")
+
+        # Force it closed since Ruby 1.8 would do an autoclose
+        # anyway, and we can close it out for now.
+        holder_io.close
+        holder_conf[:fd] = fd2
+      end
+
+      if !config
+        # Open up our own fresh file descriptor, since the desired one
+        # must now be free.
+        actual_fd = Einhorn::FD.bind(addr, port, desired_flags)
+        # This is the only case where we might change the value of
+        # desired_config[:fd], so we can't use desired_fd here.
+        desired_config[:fd] ||= actual_fd
+        if actual_fd != desired_config[:fd]
+          Einhorn::FD.dup2flags(actual_fd, desired_config[:fd])
+          IO.new(actual_fd).close
+        end
+        return
+      end
+
+      # Check on whether we need to change our existing FD number
+      if config && (current_fd = config[:fd]) != desired_fd
+        Einhorn.log_info("Renumbering #{addr}:#{port} from #{current_fd} -> #{desired_fd}")
+
+        # Actually renumber to the current value if needed
+        current_io = IO.new(current_fd)
+        Einhorn::FD.dup2flags(current_io, desired_fd)
+        current_io.close
+        config[:fd] = desired_fd
+      end
+
+      # Check on whether we need to change our existing flags
+      if config && config[:flags] != desired_flags
+        Einhorn.log_info("Changing flags on #{addr}:#{port} (#{desired_fd}) from #{config[:flags]} to #{desired_flags}")
+
+        Einhorn::FD.set_flags(desired_fd, desired_flags)
+      end
+    end
+
+    def self.load_config!
+      return unless (file = Einhorn::State.config_file)
+      load!(file)
+    end
+
+    def self.load!(filepath)
+      es = []
```

Comment:
This is a confusing variable name. It took me several reads to parse it as a plural of "e"